pkg/cache/cache.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Bartłomiej Płotka @bwplotka
+// Licensed under the Apache License 2.0.
+
+package cache
+
+import (
+	"database/sql"
+	"net/url"
+	"sync"
+	"time"
+
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/pkg/errors"
+)
+
+const (
+	driverName = "sqlite3"
+)
+
+// Storage implements a SQLite3 caching backend for Colly.
+type Storage struct {
+	// Sqlite filename.
+	Filename string
+	// Duration till which a link is skipped.
+	Validity time.Duration
+	// Clear cache at start if true.
+	ClearCache bool
+	// Database handle.
+	dbHandle *sql.DB
+	// Mutex used for clearing cache database.
+	mu sync.RWMutex
+}
+
+// Init initializes cache database.
+func (s *Storage) Init() error {
+	// Check if db exists.
+	if s.dbHandle == nil {
+		database, err := sql.Open(driverName, s.Filename)
+		if err != nil {
+			return errors.Wrap(err, "unable to open cache database file")
+		}
+
+		err = database.Ping()
+		if err != nil {
+			return errors.Wrap(err, "verify connection to cache database")
+		}
+		s.dbHandle = database
+	}
+	if s.ClearCache {
+		err := s.Clear()
+		if err != nil {
+			return err
+		}
+	}
+	// Create db with index.
+	statement, err := s.dbHandle.Prepare("CREATE TABLE IF NOT EXISTS visited (id INTEGER PRIMARY KEY, requestID INTEGER, visited INT, timestamp DATETIME)")
+	if err != nil {
+		return err
+	}
+	_, err = statement.Exec()
+	if err != nil {
+		return err
+	}
+	statement, err = s.dbHandle.Prepare("CREATE INDEX IF NOT EXISTS idx_visited ON visited (requestID)")
+	if err != nil {
+		return err
+	}
+	_, err = statement.Exec()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Clear removes all entries from cache.
+func (s *Storage) Clear() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	statement, err := s.dbHandle.Prepare("DROP TABLE visited")
+	if err != nil {
+		return err
+	}
+	_, err = statement.Exec()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close cache database.
+func (s *Storage) Close() error {
+	err := s.dbHandle.Close()
+	return err
+}
+
+// Visited inserts new URL to cache database.
+func (s *Storage) Visited(requestID uint64) error {
+	// If particular URL is already inserted, then delete.
+	// Visit method will only be called if validity expires for a URL or in case of a new URL.
```

Comment:
I would not assume the usage. Focus on Storage API. What would you expect from `Visited`. I would only expect... No idea, `visited` is the wrong name even as it suggests a bool return if something was visited. If that means just to `visit` something then we should probably bump the timestamp, nothing else, no?

Reply:
Yes, `Visited` seems like the wrong name for this but I checked this [here](https://github.com/gocolly/colly/blob/master/colly.go#L755) and [here](https://github.com/velebak/colly-sqlite3-storage/blob/master/colly/sqlite3/sqlite3.go#L111), so seems like it actually inserts it into storage. So had to do it this way.